[opencl] Update to support OpenCL 3.0

umar456 opened the pull request:
Updates OpenCL library to support OpenCL 3.0

- #### What does your PR fix?
  Fixes #23254 

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  all: Yes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

umar456 committed:
Update OpenCL to 3.0

Cheney-W commented on ports/opencl/vcpkg.json:
```diff
@@ -1,8 +1,18 @@
   "name": "opencl",
-  "version-string": "2.2",
-  "port-version": 8,
+  "version": "3.0.12",
```
"version": "3.0.12"

This version does not actually exist in the Github repository: `KhronosGroup/OpenCL-Headers` or `KhronosGroup/OpenCL-CLHPP` or `KhronosGroup/OpenCL-ICD-Loader`. 
According to above REF values, you should use `"version-string": "2022.09.30"`.

umar456 replied:
@Cheney-W OpenCL specification follows semantic versioning. The projects that create the headers for the spec seem to be following date based tags to specify releases.

As a user of OpenCL I think the version of the package should be aligned with the specification rather than the header release date. It aligns with the features that the specification offers. I understand why you would want the version string to match the tags for the release but I think that is less useful for the end user. These values aren't used when checking the project anyway so it makes little sense to set it to the date.

BillyONeal replied:
> OpenCL specification follows semantic versioning.

This isn't a release of the OpenCL specification, it's a release of `ChronosGroup/OpenCL-ICD-Loader`. Their releases follow dates, so I think the port should also have dates.

For example, if there's a bug in one of the headers that requires a specific minor update, the user needs to be able to say that in a reasonable way in a `"version>="` or an `override`.

Would it be possible to split this into separate ports corresponding to each of the upstream releases, which can use `version-date` like upstream is doing, and effectively make `opencl` an empty port that depends on those? That way we are referring to upstream using the version scheme they are publishing but also giving someone who just wants "opencl" something closer to opencl specification version numbers...

BillyONeal replied:
We could also ask the upstream maintainer(s) to document explicitly that referring to these releases by OpenCL version numbers makes sense.

umar456 replied:
OpenCL-Headers and OpenCL-ICD-Loader are typically bundled together when compute SDKs are installed. The OpenCL-Headers needs to be cloned to build the ICD(https://github.com/KhronosGroup/OpenCL-ICD-Loader#example-build). If this was part of OS package manager where runtime dependencies are separated from development dependencies, then it would make sense to separate the two. It is my understanding that VCPKG is targeted for development workflows so development files should be included together.

OpenCL specification, and these projects are maintained by the Khronos group. Their decisions are based on a group of companies that work together to create the standard. Reaching out to the maintainers may not yield a timely response. 

The releases reference the spec that the header and ICD follows in each release notes. Here is the entire note of the three repos

[OpenCL-Headers](https://github.com/KhronosGroup/OpenCL-Headers):
```
Synchronize with OpenCL v3.0.12 specification release (with OpenCL-SDK fix).
```

[OpenCL-ICD-Loader](https://github.com/KhronosGroup/OpenCL-ICD-Loader)
```
Synchronize with OpenCL v3.0.12 specification release (with OpenCL-SDK fix).
```

[OpenCL-CLHPP](https://github.com/KhronosGroup/OpenCL-CLHPP)
```
Synchronize with OpenCL v3.0.12 specification release (with OpenCL-SDK fix).
```

I understand the concern about bugs I think that can be managed via the `port-version`. I feel the version should follow the spec. This is what most users of the library reference when using the library.

dg0yt replied:
This is really tricky... But don't let us reinvent the wheel. How is it handled elsewhere?
Translationg from https://repology.org/project/opencl-headers/information:
 - Almost nobody else uses major.minor.patch `version` style  (Alpine Linux 3.16, DPorts, FreeBSD Ports).
 - Some use `version-date` style (Alpine Linux >= 3.17, Arch, ConanCenter, Gentoo, Manjaro, MSYS2, Slackware).
 - Most use either a pure major.minor `version` style, or major.minor+date variation (incl. vcpkg).
 
So IMO the version style should simply remain major.minor (now `3.0`), possibly extended by the date (`3.0+20221130`).

umar456 replied:
I am fine with `3.0`. I was under the impression the patch version was optional when specifying a restriction on the version in the manifest file. I didn't think it would limit utility of the package by adding a patch number.

dg0yt replied:
I didn't want to imply that the change to major.minor.patch imposes a drawback. But it somehow seems to add more confusion than staying with major.minor.

umar456 replied:
@Cheney-W @BillyONeal @dg0yt What is the consensus on this issue?

The options are:
- 3.0.12: The spec follows this version
- 3.0: Most users/repos expect this versioning
- v2022.09.30: The projects use this versioning when building the headers.

I personally prefer 3.0.12. but I can change it to whatever you decide.

BillyONeal replied:
It does not look like upstream has a reasonable or understandable statement here, for example https://repology.org/project/opencl-headers/versions shows different packaging systems choosing all of the above.

I'm still in favor of matching the github release exactly. If upstream's users find that confusing, upstream should release versions that users don't find confusing. It doesn't seem right for vcpkg to be making up a version which is not the version upstream published.

If upstream would make a statement in terms of how they expect these bits to be versioned that would be helpful.

Cheney-W replied:
@umar456 Please use "version-string": "2022.09.30".

umar456 commented:
Fixed the license type and changed version to version-semver which more accurately aligns with the OpenCL specification.

I can change it to date based versioning if @Cheney-W disagrees with my reasoning behind my use of the semantic versioning argument I used in my last comment.

umar456 committed:
Update OpenCL versions baseline file

umar456 commented:
Addressed all feedback.
